nixos/grub: support initrd secrets

lopsided98 opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Long delayed second try at #38263.

I have been running this for a long time and it has worked well for me. @emilazy and @Kloenk could you try and see if this works for you? @mdorman If you can clarify the issues you ran into, I'll try to address them.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

lopsided98 committed:
nixos/grub: support initrd secrets

emilazy commented on nixos/modules/system/boot/loader/grub/grub.nix:
```diff
@@ -292,19 +293,6 @@ in
         '';
       };
 
-      extraInitrd = mkOption {
-        type = types.nullOr types.path;
-        default = null;
-        example = "/boot/extra_initramfs.gz";
-        description = ''
-          The path to a second initramfs to be supplied to the kernel.
-          This ramfs will not be copied to the store, so that it can
-          contain secrets such as LUKS keyfiles or ssh keys.
-          This implies that rolling back to a previous configuration
-          won't rollback the state of this file.
-        '';
-      };
-
```
Looks good to me; I'll try to test soon. A nitpick: if we're going to remove the `extraInitrd` option then it'd be good to add a `mkRemovedOptionModule` so people get error messages that point them to the right place and we can avoid repeating #41608; e.g.:

```nix
{
  imports = [
    (mkRemovedOptionModule [ "boot" "loader" "grub" "extraInitrd" ] ''
      ...some brief pointer to the initrd secrets options/docs, maybe with conversion example...
    '')
  ];
}
```



Another option would be to just leave in support for `extraInitrd` for now to give people time to migrate; I don't have strong feelings about this since I don't use GRUB, and I guess breaking compatibility saves adding deprecation warnings. But the `zcat | cpio` dance to extract pre-baked secrets in this format is a little arcane, so it'd be nice to document it.

symphorien replied:
Is this option fundamentally incompatible with the new mechanism ?
Also: this requires release notes, no just a warning, as it makes systems unbootable.

lopsided98 replied:
It would be possible to keep the `extraInitrd` option, but I wanted to limit complexity and I don't know of any use case for it other than including secrets, which is far better served by the `boot.initrd.secrets` option.

I added the `mkRemovedOptionModule` warning, but I don't know how to explain the conversion process because I have never done it myself.

emilazy replied:
Something like:

```
To migrate to the new initrd secrets system, extract the extraInitrd
archive into your main filesystem:

  # zcat /boot/extra_initramfs.gz | cpio -idvmD /etc/secrets/initrd
  /path/to/secret1
  /path/to/secret2
  # rm /boot/extra_initramfs.gz

then replace boot.loader.grub.extraInitrd with boot.initrd.secrets:

  boot.initrd.secrets = {
    "/path/to/secret1" = "/etc/secrets/initrd/path/to/secret1";
    "/path/to/secret2" = "/etc/secrets/initrd/path/to/secret2";
  };

See the boot.initrd.secrets option documentation for more information.
```

Mic92 commented on nixos/modules/system/boot/loader/grub/install-grub.pl:
```diff
@@ -363,9 +356,23 @@ sub addEntry {
+    # Include second initrd with secrets
+    if ($initrdSecrets) {
+      # Get last element of path
+      $initrd =~ /\/([^\/]+)$/;
```
I think perl also has a proper basename function in its stdlib.

Kloenk commented on nixos/modules/system/boot/loader/grub/install-grub.pl:
```diff
@@ -363,9 +356,23 @@ sub addEntry {
+      umask 0137;
+      my $initrdSecretsPathTemp = File::Temp::mktemp("$initrdSecretsPath.XXXXXXXX");
+      system("$path/append-initrd-secrets", $initrdSecretsPathTemp) == 0 or die "failed to create initrd secrets\n";
+      rename $initrdSecretsPathTemp, $initrdSecretsPath or die "failed to move initrd secrets into place\n";
```
I'm running into this die statement, but I'm not sure what I am doing wrong

Kloenk commented:
I always get an failed to move initrd secrets into place error. This even happens if I have no initrd secrets on that host.

lopsided98 committed:
nixos/stage-1: make boot.initrd.secrets appear in the manual

lopsided98 commented:
@Kloenk That should be fixed now. 

I also fixed another issue that would happen when `copyKernels == false`, but I still haven't tested that configuration. My only system that uses GRUB has an encrypted rootfs, and I haven't figured out how to test GRUB in a NixOS VM, because they normally don't use a bootloader.

I also removed `internal = true` from `boot.secrets.initrd`, which was the reason it didn't appear in the manual and @mdorman thought it was undocumented.

Kloenk commented:
It still fails. I don't know what the problem is, and I don't have an clue on where to start debugging.

The /boot/kernels directory contians an `<hash>-initrd-secrets` file, but this initrd is use by 3 generations bevore.

The log output I get from the `nixos-rebuild switch ...` command is:
```
updating GRUB 2 menu...
4 blocks
4 blocks
4 blocks
4 blocks
4 blocks
4 blocks
failed to move initrd secrets into place
```

The non initrd-secrets machine is still building, will post if it also fails tomorrow.

emilazy commented:
Are you sure you aren't using a module that uses the initrd secrets functionality under the hood (e.g. the initrd SSH module)? `nixos-option boot.initrd.secrets` might reveal something.

lopsided98 commented:
Oh, I see now. The current code assumes that all generations either do or don't have initrd secrets enabled, which obviously isn't the case when you first enable the option.

emilazy commented:
One disadvantage of removing support for `extraInitrd` immediately is that I guess it will immediately break boots of all old generations that use it as soon as you perform a rebuild, so you can't roll back if you mess up migration to initrd secrets.

Kloenk commented:
> Are you sure you aren't using a module that uses the initrd secrets functionality under the hood (e.g. the initrd SSH module)? `nixos-option boot.initrd.secrets` might reveal something.

The initrd ssh module is the only secret I use. On my local PC (at the moment building krita) I don't have initrd ssh. On my server I have ssh, and there it is failing

lopsided98 commented:
> One disadvantage of removing support for `extraInitrd` immediately is that I guess it will immediately break boots of all old generations that use it as soon as you perform a rebuild, so you can't roll back if you mess up migration to initrd secrets.

The existing implementation already has that problem. The current generation's `extraInitrd` is applied to all generations. If you disable `extraInitrd` in the current generation, it gets disabled in all generations.

lopsided98 commented:
I rewrote the implementation to be more similar to [how systemd-boot does it](https://github.com/NixOS/nixpkgs/blob/f5789139051c26ceb2eaf3ebaa0f8a3aa3424fdb/nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py#L94). It should now work correctly when only some of the generations have secrets.

lopsided98 commented:
I have fixed all the known issues, but I think it still needs more testing, particularly on systems with the a single boot/root partition.

Kloenk commented:
I finaly got around to deploy it on my initrd-ssh-server machines. It works there, but only after I did a `nix-collect-garbage -d`, because of old generations of my systems, for whom it tried to build initrd-secrets for

emilazy commented:
Seems like an instance of #85563. Was the error just that the `extraInitrd` file no longer existed, or was the initrd secrets code somehow running on the old generations and failing? If the former, we should maybe just ditch the advice to `rm` it to save people problems rolling back.

Kloenk commented:
As far as i understood it, it tried to create an initrd for a previos generation. The config of this generation looked for the initrd secret on another path, and so it failed

lopsided98 commented:
Like @emilazy said, that sounds like an issue with the current initrd secrets implementation, not specific to GRUB.

> If the former, we should maybe just ditch the advice to `rm` it to save people problems rolling back.

The documentation doesn't currently suggest that users delete the initrd file. I think rollbacks (ie. `nixos-rebuild switch --rollback`) will still work as long as the initrd file still exists, depending on whether `nixos-rebuild` runs the bootloader install script in that case. If you directly boot into an older generation, the extra initrd will be missing. There is really no way to fix this, because there is no way to know what initrd file was configured for an older generation.

Kloenk commented:
So it seems like everything is working as intendet. Maybe write something about old profiles in the release note, so nobody runs into the same problem as me?

lopsided98 commented:
I added some clarification to the changelog, describing the issue I mentioned in my previous comment. It doesn't really talk about the issue @Kloenk ran into, because that isn't related to GRUB or this PR specifically.

Kloenk commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -441,6 +441,20 @@ systemd.services.nginx.serviceConfig.ReadWritePaths = [ "/var/www" ];
+     initrd from the specified files, rather than using a manually created
+     initrd file.
+
+     Due to an existing bug with <option>boot.loader.grub.extraInitrd</option>,
```
I think this is enough so if somebody has an error like mine they have a point to start searching

kloenk reviewed (approved):
Still works on my servers. So I think it is good to go

Mic92 commented on nixos/modules/system/boot/loader/grub/install-grub.pl:
```diff
@@ -363,9 +355,30 @@ sub addEntry {
+      # Make sure initrd is not world readable (won't work if /boot is FAT)
+      umask 0137;
+      my $initrdSecretsPathTemp = File::Temp::mktemp("$initrdSecretsPath.XXXXXXXX");
+      system("$path/append-initrd-secrets", $initrdSecretsPathTemp) == 0 or die "failed to create initrd secrets\n";
```
Since we might deal with fat, it would be great to also use something like syncfs here in one way or the other. https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py#L231

It might be enough to do one syncfs at the end of the grub builder for the EFI partition.
I will open a new PR for that.

Mic92 replied:
Ok. I think, grub already takes care of this.

Mic92 commented:
Tested on my server as well.

ornxk commented:
it doesn't seem to rebuild the secrets file if the contents of the secrets change
in my case, i had accidentally set a secret to the wrong file, but upon fixing it and rebuilding, the secrets file remained unchanged

lopsided98 commented:
I'm not sure how that could happen. The initrds for all generations should be generated each time you activate a new configuration. Could you provide more details? Did you change the path to the secret file between the generations, or just its contents?

CRTified commented:
The second commit also causes the `initrd-network-ssh` test to fail. It looks like the secret is not copied into the initramfs.

lopsided98 commented:
I'm not totally sure yet, but I think this is a bug in the test. Before this PR, GRUB did not support initrd secrets, so the secrets were copied to the Nix store. Now, they need to be copied by the bootloader install script, which does not run because the test does not use a bootloader.

Setting `virtualisation.useBootLoader` might be the right solution, but it causes the build to fail for me.

lopsided98 commented:
See #91744

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/full-encrypted-nixos-system-on-legacy-boot-with-secrets-and-remote-unlock-for-unstable-20-03/8279/1
